video: display: misc changes - draft

ngphibang opened the pull request:
This **draft** PR contains various changes to the video subsystem and some for display that are already or will be presented in other PRs.

The purpose of collecting them here is to serve as a dependency for the [libMP PR](https://github.com/zephyrproject-rtos/zephyr/pull/98514)

ngphibang committed:
drivers: display: mcux_elcdif: Decouple the PxP from the elcdif driver

PxP is a distinct HW IP for image processing which is independent of the
display controller. Decouple it from the elcdif driver.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
video: controls: Add VIDEO_CID_ROTATE

Add VIDEO_CID_ROTATE which is needed for some m2m devices.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: dma: Drop NXP's PxP driver

Drop the old dma PxP driver to favor the new one in v4z m2m category.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Add PxP m2m driver

The NXP's PxP is a 2D pixel engine which is capable to do some image
transformation such as rotation, flipping, color conversion, etc.

Make a driver for it in the video m2m category.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Request buffers instead of allocating

Replace video_buffer_alloc() and video_buffer_aligned_alloc() APIs with the
new video_request_buffers() API. With this,

- We can request multiple buffers at once.

- Application does not need to keep the empty allocated buffer pointer
  as it can enqueue the buffer via its index.

- Application may allocate its own buffers and provide them to the
  framework.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Move buffer management stuffs into a separate file

Move buffer management stuffs into video_buffer.c

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Enqueue and track internal buffers via index

Currently, the video_enqueue() API enqueues the whole external video
buffer container structure from the application and the driver stores
this container in its FIFO queue.

While it works in simple applications where the enqueued video_buffer
container persists for the whole program lifecycle, it does not work in
situations where we cannot keep this container, e.g. enqueuing a buffer
inside a function, the local variable will be destroyed when the function
returns and hence the buffer is no longer valid.

Video buffers can be tracked via their indices in the buffer pool.
Set the index field when buffers are allocated and enqueue the internal
buffer rather than the external one to fix the issue.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: video_common: Drop mem_block structure

The mem_block structure and the video_block array are not needed.
Drop them.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Support user-provided buffers

Add support for buffers allocated outside the video buffer heap.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: drivers: video: Some misc fixes

Some misc fixes:

- Use uint16_t for video_buffer's index
- Don't use enum in struct definition as it will be translated to int or
  other types depending on compiler and platform. Use uint8_t to save
  some bytes.
- Other minor fixes

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: drivers: video: Use errno instead of _ASSERT in set_caps()

Return an errno instead of _ASSERT because in some cases, the propram
needs to be continued even if a function failed. Furthermore, _ASSERT
shouldn't be used in production.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Add buffer alignment to caps

Add a field to video_caps to expose buffer alignment requirement.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: display: Add XRGB format

Add PIXEL_FORMAT_XRGB_8888 and the corresponding panel format.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: display: mcux_elcdif: Fix ARGB/XRGB format issue

The eLCDIF supports XRGB format but the ARGB format was used instead.
Fix it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: drivers: video: Add YVU multi-planar formats

Add VIDEO_PIX_FMT_YVU422M and VIDEO_PIX_FMT_YVU420M formats

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
include: drivers: video: Introduce transform_caps API

For m2m video device, the get_caps() API give the supported caps of both
input and output sides. However, there is currently no guaranteed way to
know the relationship between input and output caps.

Introduce a new transform_caps() API that help to transform a video
format cap from one end to the other end of a m2m device.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: ov5640: Expose the high framerates first

Expose the high framerates first so that it can be listed first and
chosen as the default frame rate in some applications.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
Update:

- Fix pxp's transform_caps()

danieldegrasse commented on include/zephyr/drivers/video.h:
```diff
@@ -46,9 +46,9 @@ struct video_control;
+	VIDEO_BUF_TYPE_INPUT = 1,
 	/** output buffer type */
-	VIDEO_BUF_TYPE_OUTPUT,
+	VIDEO_BUF_TYPE_OUTPUT = 2,
```
If we are going to limit this type to 1 byte, we should likely set a `VIDEO_BUF_TYPE_MAX` in this enum as well to prevent developers from adding too many constants in the future

ngphibang replied:
Yes, we can do this as an extremely careful prevention. But as other enums in video.h (e.g. video_frmival_type, etc.), I think it is very difficult to add another item into this, we only see these two items are needed to support m2m, otherwise if someone add more type, he/she must have a very strong reason so the prevention will not be needed, I think

danieldegrasse reviewed (commented):
I'm aware this is a draft- just thought I'd add some initial comments:
- I agree that making the video API work with buffer requests rather than allowing the application to allocate buffers at a size it chooses is likely better, but this will be a stable API change. The application should already be able to determine the size needed for each video buffer based on the format set, right? Or is there a reason I'm missing that we need this API change
- I like the rework of the PXP into a M2M driver- we might want a sample of how to use this though, I know that customers have requested this in the past. Another issue to figure out is how to integrate the PXP into LVGL's processing, right? Or would we expect an LVGL user to use the LVGL 2D accelerator support for the PXP rather than the M2M driver

ngphibang committed:
dts: arm: nxp: nxp_rt11xx: Choose pxp as video transform node

Define pxp as a chosen video transform node.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
> I like the rework of the PXP into a M2M driver- we might want a sample of how to use this though, I know that customers have requested this in the past. Another issue to figure out is how to integrate the PXP into LVGL's processing, right? Or would we expect an LVGL user to use the LVGL 2D accelerator support for the PXP rather than the M2M driver

Hi @danieldegrasse ,

Sorry for the late reply. Yes, the difficult part is not on the driver but on how to integrate it into different Zephyr samples as a separate m2m driver (where in the current upstream it is coupled with the LCDIF to ease the work).

For the video capture sample, I leveraged the sample to achieve this as can be seen here https://github.com/zephyrproject-rtos/zephyr/pull/92366

However, for other samples (LVGL, display), it needs to be done similarly but will be more difficult as it requires a generic abstraction that must work for all other boards not only NXP's.

For LVGL in particular, I think we should not pass by LVGL as LVGL bypasses Zephyr layer and goes directly to the low level driver, this is not good IMHO and can cause some conflicts as mentionned partly in this [comment](https://github.com/zephyrproject-rtos/zephyr/pull/98514#issuecomment-3467415879).

With [libMP](https://github.com/zephyrproject-rtos/zephyr/pull/98514#issuecomment-3467415879), hopefully the PxP integration into Zephyr samples will be easier.

Commit:
video: stm32-dcmipp: dcmipp requires 16bytes aligned buffers

Add the buf_align entry in caps to indicate buffer alignment
constraints.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

ngphibang committed:
drivers: display: st7796s: Fix get_capabilities

The supported pixel formats are the most important thing in
get_capabilities() but not reported. Fix it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: display: st7796s: Add empty set_pixel_format

Add an empt set_pixel_format() implementation

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -135,6 +135,8 @@ struct video_caps {
 	 */
 	uint8_t min_vbuf_count;
+	/** requirement on the buffer alignment, in bytes */
+	size_t buf_align;
```
Should the convention be `0` for no alignment requirement, which is the same as `1` byte alignment?

Otherwise, `0` means misconfigured alignment and with this commit, all drivers become invalid?

josuah commented on include/zephyr/drivers/video.h:
```diff
@@ -741,6 +754,48 @@ static inline int video_get_caps(const struct device *dev, struct video_caps *ca
+static inline int video_transform_cap(const struct device *const dev,
+				      const struct video_format_cap *const cap,
+				      struct video_format_cap *const res_cap,
+				      enum video_buf_type type, uint16_t ind)
```
As a summary of the discussion on Discord:

The video devices that convert data (i.e. video compression, scaling...) will have both input formats capabilities and output formats capabilities. But this does not mean that every input will work with every output. For instance, a *down*scaler will not allow to reach higher resolution than the source. So some extra querying is needed.

In order to tell "for this input format, what output format can I use" this new API is introduced, letting the application probe one by one every format it is interested in, and see what the format at the other end would be for it.

KATE-WANG-NXP committed:
[DNM]drivers: video: sw_generator: Add JPEG and PNG format support

Add support for compressed video formats (JPEG and PNG) in the software
video generator driver. This enables testing of compressed video formats
without requiring actual hardware.
The compressed formats only support a fixed 320x160 resolution and
provide two frames (normal and horizontally flipped) for testing.

Signed-off-by: Kate Wang <yumeng.wang@nxp.com>

ngphibang committed:
[DNM]drivers: video: sw_generator: Fix Jpeg size

Jpeg size is wrong. Fix it.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
drivers: video: Support different alignment request

Currently, buffer allocation is using the same alignment specified in
the CONFIG_VIDEO_BUFFER_POOL_ALIGN.

Add support for a different alignment in each buffer request.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang committed:
samples: video: capture: Support different alignment request

Add support for a different alignment in each buffer request.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

josuah commented:
Let us know when you think is a good time to start walking through this content.

It is possible that splitting the video-related and display-related parts will help.

Glad to help with what I can as time permits.

ngphibang commented:
> Let us know when you think is a good time to start walking through this content.

It is possible that splitting the video-related and display-related parts will help.

Thanks @josuah. This is a draft PR as a dependency for libMP PR. I already split it into several PRs but didn't have time to address the review yet. Will try to do in the next days, sorry for the delay.

ngphibang committed:
[DNM]include: drivers: video: Add PNG fourcc format

This is to quickly add a fake fourcc format for PNG. We don't have such
a fourcc as PNG is a file format.

Signed-off-by: Phi Bang Nguyen <phibang.nguyen@nxp.com>

ngphibang commented:
This PR is closed as stale and I forgot to re-open it then I pushed force to the branch when it is closed. Now I cannot re-open it anymore. Do you know how to proceed ?
